src/checksum/arm64/crc32-arm64.c

```diff
@@ -59,30 +60,46 @@ static uint32_t
 __wt_checksum_with_seed_hw(uint32_t seed, const void *chunk, size_t len)
 {
     uint32_t crc;
-    size_t nqwords;
+    uint64_t uint64_v;
+    uint32_t uint32_v;
+    uint16_t uint16_v;
     const uint8_t *p;
-    const uint64_t *p64;
 
     crc = ~seed;
 
-    /* Checksum one byte at a time to the first 4B boundary. */
-    for (p = chunk; ((uintptr_t)p & (sizeof(uint32_t) - 1)) != 0 && len > 0; ++p, --len) {
-        CRC32CB(crc, *p);
+    p = (const uint8_t *)chunk;
+    /*
+     * Checksum in 16B chunks. Do two loads at a time as the looping logic generated by the compiler
+     * is faster.
+     */
+    for (; len >= 2 * sizeof(uint64_t); p += 2 * sizeof(uint64_t), len -= 2 * sizeof(uint64_t)) {
+        memcpy(&uint64_v, p, sizeof(uint64_t));
+        CRC32CX(crc, uint64_v);
+        memcpy(&uint64_v, p + sizeof(uint64_t), sizeof(uint64_t));
+        CRC32CX(crc, uint64_v);
     }
 
-    p64 = (const uint64_t *)p;
-    /* Checksum in 8B chunks. */
-    for (nqwords = len / sizeof(uint64_t); nqwords; nqwords--) {
-        CRC32CX(crc, *p64);
-        p64++;
+    if (len & sizeof(uint64_t)) {
```

Comment:
Is the bitwise and part of the performance story? I think the intent would be clearer if we used `>=`.

Reply:
Not really. Changed as suggested.

Reply:
Sorry I didn't mean it like that. I meant like `len >= sizeof(uint64_t)`.

Reply:
For perf, I don't think they make any difference.

Reply:
It *was* part of the perf story, for two reasons:

1) arm64 has single instruction conditional branches for testing if a register is (not) zero (`cbz` and `cbnz`) and testing is a specific bit in a register is (not) zero (`tbz` and `tbnz`). Testing for >= a non-zero value requires a comparison to set flags and a flag-based branch. And that is in addition to modifying the `len` which requires an additional unnecessary instruction. Neoverse N1 (grav2) and V1 (grav3) don't seem to do x86-style macro-op fusion between the comparison and the branch, although Neoverse V2 does (judging just based on their published optimizations guides for all 3).
2) as originally written, those branches don't depend on each other. Whether you enter the `len & 4` block had no dependency on whether you entered the `len & 8` block and so on. As modified, each branch condition depends on whether you entered the prior branch because each block both modifies and branches on `len`. That means that they can't be evaluated in parallel by the CPU.

compare:

```asm
  401b3c:       36180061        tbz     w1, #3, 401b48 <hashBytes(void*, unsigned long)+0x48>
  401b40:       f8408460        ldr     x0, [x3], #8
  401b44:       9ac05c42        crc32cx w2, w2, x0
  401b48:       36100061        tbz     w1, #2, 401b54 <hashBytes(void*, unsigned long)+0x54>
  401b4c:       b8404460        ldr     w0, [x3], #4
  401b50:       1ac05842        crc32cw w2, w2, w0
  401b54:       36080061        tbz     w1, #1, 401b60 <hashBytes(void*, unsigned long)+0x60>
  401b58:       78402460        ldrh    w0, [x3], #2
  401b5c:       1ac05442        crc32ch w2, w2, w0
  401b60:       36000061        tbz     w1, #0, 401b6c <hashBytes(void*, unsigned long)+0x6c>
  401b64:       39400060        ldrb    w0, [x3]
  401b68:       1ac05042        crc32cb w2, w2, w0
  401b6c:       2a2203e0        mvn     w0, w2
  401b70:       d65f03c0        ret
```

```asm
  401aa0:       f1001c3f        cmp     x1, #0x7
  401aa4:       54000089        b.ls    401ab4 <__wt_checksum(void const*, unsigned long)+0x54>  // b.plast
  401aa8:       d1002021        sub     x1, x1, #0x8
  401aac:       f8408460        ldr     x0, [x3], #8
  401ab0:       9ac05c42        crc32cx w2, w2, x0
  401ab4:       f1000c3f        cmp     x1, #0x3
  401ab8:       54000089        b.ls    401ac8 <__wt_checksum(void const*, unsigned long)+0x68>  // b.plast
  401abc:       d1001021        sub     x1, x1, #0x4
  401ac0:       b8404460        ldr     w0, [x3], #4
  401ac4:       1ac05842        crc32cw w2, w2, w0
  401ac8:       f100043f        cmp     x1, #0x1
  401acc:       54000089        b.ls    401adc <__wt_checksum(void const*, unsigned long)+0x7c>  // b.plast
  401ad0:       d1000821        sub     x1, x1, #0x2
  401ad4:       78402460        ldrh    w0, [x3], #2
  401ad8:       1ac05442        crc32ch w2, w2, w0
  401adc:       b4000061        cbz     x1, 401ae8 <__wt_checksum(void const*, unsigned long)+0x88>
  401ae0:       39400060        ldrb    w0, [x3]
  401ae4:       1ac05042        crc32cb w2, w2, w0
  401ae8:       2a2203e0        mvn     w0, w2
  401aec:       d65f03c0        ret
```

It was a bit difficult to test this meaningfully because this effects tail perf in a way that a branch predictor can trivially predict if you use a hardcoded size or pattern. I ended up using an array of 2000 random lens from 0-15, and looping over that 1 million times. In that (admittedly artificial) scenario, using the bit-branching was about 10% faster. 

While I generally hate argument from authority, this does seem to be how basically everyone does it: [linux asm](https://github.com/torvalds/linux/blob/1b3dce8b8ab30bb9f1401acefa08a47dc46f0813/arch/arm64/lib/crc32.S#L69-L84), [older linux C code](https://elixir.bootlin.com/linux/v4.6/source/arch/arm64/crypto/crc32-arm64.c#L74), [fio](https://github.com/axboe/fio/blob/06c40418f97811092c0aece1760487400bcdd506/crc/crc32c-arm64.c#L83-L94), [google crc32c](https://github.com/google/crc32c/blob/21fc8ef30415a635e7351ffa0e5d5367943d4a94/src/crc32c_arm64.cc#L104-L116), [ceph](https://github.com/ceph/ceph/blob/9085e776af8cb21c6687f23088bfcedd8fd44a65/src/common/crc32c_aarch64.c#L263-L271). Some of these use a more complicated version of the bulk loop using `vmull_p64`, but I tried the fio version and it was slower than our simple 16-byte-at-a-time version on graviton2, even on 1MB blocks, so it probably isn't a good idea for us to adopt that. Might be worth re-evaluating if we ever move to grav3 through since it has more parallel execution resources.

Reply:
Thanks @RedBeard0531. I reverted the change.

Reply:
Thanks @RedBeard0531 . 10% on a microbenchmark like that is marginal when trading off readability, but it's Chenhao's call and I can totally see the justification. This still LGTM.